src/background-view.cpp

```diff
@@ -0,0 +1,355 @@
+                 */
+                view->set_geometry(o->get_relative_geometry());
+
+                /* Set it as the background */
+                o->workspace->add_view(view, wf::LAYER_BACKGROUND);
+
+                /* Make it show on all workspaces */
+                view->role = wf::VIEW_ROLE_DESKTOP_ENVIRONMENT;
+
+                procs[o].view = view;
+
+                break;
+            }
+        }
+    }};
+
+    std::string add_arg_if_not_empty(std::string in)
+    {
+        if (!in.empty())
+        {
+            return " \"" + in + "\"";
+        }
+        else
+        {
+            return in;
+        }
+    }
+
+    /* The following functions borrowed from weston */
+    int os_fd_set_cloexec(int fd)
+    {
+        long flags;
+
+        if (fd == -1)
+        {
+            return -1;
+        }
+
+        flags = fcntl(fd, F_GETFD);
+        if (flags == -1)
+        {
+            return -1;
+        }
+
+        if (fcntl(fd, F_SETFD, flags | FD_CLOEXEC) == -1)
+        {
+            return -1;
+        }
+
+        return 0;
+    }
+
+    int set_cloexec_or_close(int fd)
+    {
+        if (os_fd_set_cloexec(fd) != 0)
+        {
+            close(fd);
+            return -1;
+        }
+        return fd;
+    }
+
+    int os_socketpair_cloexec(int domain, int type, int protocol, int *sv)
+    {
+        int ret;
+
+#ifdef SOCK_CLOEXEC
+        ret = socketpair(domain, type | SOCK_CLOEXEC, protocol, sv);
+        if (ret == 0 || errno != EINVAL)
+        {
+            return ret;
+        }
+#endif
+
+        ret = socketpair(domain, type, protocol, sv);
+        if (ret < 0)
+        {
+            return ret;
+        }
+
+        sv[0] = set_cloexec_or_close(sv[0]);
+        sv[1] = set_cloexec_or_close(sv[1]);
+
+        if (sv[0] != -1 && sv[1] != -1)
+        {
+            return 0;
+        }
+
+        close(sv[0]);
+        close(sv[1]);
+        return -1;
+    }
+
+    static int sigchld_handler(int signal_number, void *data)
+    {
+        std::map<wf::output_t *, struct process> *procs = (std::map<wf::output_t *, struct process> *) data;
+        pid_t pid;
+        int status;
+
+        while ((pid = waitpid(-1, &status, WNOHANG)) > 0) {
+            for (auto& p : *procs)
+            {
+                if (p.second.pid == pid)
+                {
+                    p.second.view = nullptr;
+                    break;
+                }
+            }
+        }
+
+        if (pid < 0 && errno != ECHILD)
+        {
+            LOGE("waitpid error: ", strerror(errno));
+        }
+
+        return 1;
+    }
+
+    void child_client_exec(int sockfd, const char *command)
+    {
+        int clientfd;
+        char s[32];
+        sigset_t allsigs;
+
+        /* do not give our signal mask to the new process */
+        sigfillset(&allsigs);
+        sigprocmask(SIG_UNBLOCK, &allsigs, NULL);
+
+        /* Launch clients as the user. Do not launch clients with wrong euid. */
```

Comment:
I think we drop the root privileges as soon as we load the backend (which is before plugins are loaded).

At the point where this plugin is loaded, we are regular user so no need for this.